libeos-updater-util/avahi-service-file.c

```diff
@@ -38,65 +41,131 @@ const gchar * const eos_avahi_v1_ostree_path = "eos_ostree_path";
+  while (g_variant_iter_loop (&iter, "(&sv)", &txt_key, &txt_value))
     {
-      g_autofree gchar *record_escaped = g_markup_escape_text (txt_records[idx], -1);
-      g_string_append_printf (str, "    <txt-record>%s</txt-record>\n",
-                              record_escaped);
+      const GVariantType *txt_value_type = g_variant_get_type (txt_value);
+      gboolean txt_value_is_binary = g_variant_type_equal (txt_value_type, G_VARIANT_TYPE_BYTESTRING);
+      gboolean txt_value_is_text = g_variant_type_equal (txt_value_type, G_VARIANT_TYPE_STRING);
+      g_autofree gchar *txt_key_escaped = g_markup_escape_text (txt_key, -1);
+
+      g_assert (txt_value_is_binary || txt_value_is_text);
+
+      if (txt_value_is_binary)
+        {
+          gsize txt_value_data_length;
+          gconstpointer txt_value_data = g_variant_get_fixed_array (txt_value,
+                                                                    &txt_value_data_length,
+                                                                    1);
+          g_autofree gchar *txt_value_encoded = NULL;
+          g_autofree gchar *txt_value_escaped = NULL;
+
+          if (txt_value_data_length + strlen (txt_key) + 1 > 255)
+            g_warning ("The TXT record with key %s is longer than 255 bytes, it will be truncated!",
+                       txt_key);
+
+          txt_value_encoded = g_base64_encode (txt_value_data, txt_value_data_length);
+          txt_value_escaped = g_markup_escape_text (txt_value_encoded, -1);
+          g_string_append_printf (str,
+                                  "    <txt-record value-format=\"binary-base64\">%s=%s</txt-record>\n",
+                                  txt_key_escaped,
+                                  txt_value_escaped);
+        }
+      else
+        {
+          gsize txt_value_string_length;
+          const gchar *txt_value_string = g_variant_get_string (txt_value,
+                                                                &txt_value_string_length);
+          g_autofree gchar *txt_value_escaped = NULL;
+
+          if (txt_value_string_length + strlen(txt_key) + 1 > 255)
+            g_warning ("The TXT record with key %s is longer than 255 bytes, it will be truncated!",
+                       txt_key);
+          txt_value_escaped = g_markup_escape_text (txt_value_string,
+                                                    txt_value_string_length);
+          g_string_append_printf (str, "    <txt-record>%s=%s</txt-record>\n",
+                                  txt_key_escaped,
+                                  txt_value_escaped);
+        }
     }
 
+  // TODO: compute the total size of txt records and complain loudly
+  // if they are longer then certain sizes:
+  //
+  // - more than 256 bytes, because some crappy hardware did not work
+  //   properly with longer records (likely not applicable for us and
+  //   that would limit the size of the bloom filter severely)
```

Comment:
I added another option, so you can specify which size level you don't want to break.